source/common/ext_authz/ext_authz_impl.cc

```diff
@@ -0,0 +1,189 @@
+
+GrpcClientImpl::GrpcClientImpl(Grpc::AsyncClientPtr&& async_client,
+                               const Optional<std::chrono::milliseconds>& timeout)
+    : service_method_(*Protobuf::DescriptorPool::generated_pool()->FindMethodByName(
+          "envoy.service.auth.v2.Authorization.Check")),
+      async_client_(std::move(async_client)), timeout_(timeout) {}
+
+GrpcClientImpl::~GrpcClientImpl() { ASSERT(!callbacks_); }
+
+void GrpcClientImpl::cancel() {
+  ASSERT(callbacks_ != nullptr);
+  request_->cancel();
+  callbacks_ = nullptr;
+}
+
+void GrpcClientImpl::check(RequestCallbacks& callbacks,
+                           const envoy::service::auth::v2::CheckRequest& request,
+                           Tracing::Span& parent_span) {
+  ASSERT(callbacks_ == nullptr);
+  callbacks_ = &callbacks;
+
+  request_ = async_client_->send(service_method_, request, *this, parent_span, timeout_);
+}
+
+void GrpcClientImpl::onSuccess(std::unique_ptr<envoy::service::auth::v2::CheckResponse>&& response,
+                               Tracing::Span& span) {
+  CheckStatus status = CheckStatus::OK;
+  ASSERT(response->status().code() != Grpc::Status::GrpcStatus::Unknown);
+  if (response->status().code() != Grpc::Status::GrpcStatus::Ok) {
+    status = CheckStatus::Denied;
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceUnauthz);
+  } else {
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceOk);
+  }
+
+  callbacks_->onComplete(status);
+  callbacks_ = nullptr;
+}
+
+void GrpcClientImpl::onFailure(Grpc::Status::GrpcStatus status, const std::string&,
+                               Tracing::Span&) {
+  ASSERT(status != Grpc::Status::GrpcStatus::Ok);
+  UNREFERENCED_PARAMETER(status);
+  callbacks_->onComplete(CheckStatus::Error);
+  callbacks_ = nullptr;
+}
+
+void CreateCheckRequest::setAttrContextPeer(envoy::service::auth::v2::AttributeContext_Peer& peer,
+                                            const Network::Connection& connection,
+                                            const std::string& service, const bool local) {
+
+  // Set the address
+  auto addr = peer.mutable_address();
+  if (!local) {
+    Envoy::Network::Utility::addressToProtobufAddress(*connection.remoteAddress(), *addr);
+  } else {
+    Envoy::Network::Utility::addressToProtobufAddress(*connection.localAddress(), *addr);
+  }
+
+  // Set the principal
+  // Preferably the SAN from the peer's cert or
+  // Subject from the peer's cert.
+  std::string principal;
+  Ssl::Connection* ssl = const_cast<Ssl::Connection*>(connection.ssl());
+  if (ssl != nullptr) {
+    if (!local) {
+      principal = ssl->uriSanPeerCertificate();
+
+      if (principal.empty()) {
+        principal = ssl->subjectPeerCertificate();
+      }
+    } else {
+      principal = ssl->uriSanLocalCertificate();
+
+      if (principal.empty()) {
+        principal = ssl->subjectLocalCertificate();
+      }
+    }
+  }
+  peer.set_principal(principal);
```

Comment:
We probably need some format normalization here. At least document what is the format used here?